test/gtest/ucp/test_ucp_tag_mem_type.cc

```diff
@@ -0,0 +1,216 @@
+                generate_test_params_variant(ctx_params, name, test_case_name + "/" +
+                                             std::string(ucs_memory_type_names[(*iter)[0]]) +
+                                             "<->" + std::string(ucs_memory_type_names[(*iter)[1]]),
+                                             tls, count++, result);
+            }
+        }
+
+        return result;
+    }
+
+    static std::vector<std::vector<ucs_memory_type_t> > mem_type_pairs;
+
+protected:
+
+    size_t do_xfer(const void *sendbuf, void *recvbuf, size_t count,
+                   ucp_datatype_t send_dt, ucp_datatype_t recv_dt,
+                   bool expected, bool truncated, bool extended);
+
+    mem_buffer        *m_recv_mem_buf;
+    mem_buffer        *m_send_mem_buf;
+    ucs_memory_type_t m_send_mem_type;
+    ucs_memory_type_t m_recv_mem_type;
+
+private:
+
+    static const uint64_t SENDER_TAG = 0x111337;
+    static const uint64_t RECV_MASK  = 0xffff;
+    static const uint64_t RECV_TAG   = 0x1337;
+};
+
+std::vector<std::vector<ucs_memory_type_t> >
+test_ucp_tag_mem_type::mem_type_pairs = ucs::supported_mem_type_pairs();
+
+size_t test_ucp_tag_mem_type::do_xfer(const void *sendbuf, void *recvbuf,
+                                  size_t count, ucp_datatype_t send_dt,
+                                  ucp_datatype_t recv_dt, bool expected,
+                                  bool truncated, bool extended)
+{
+    size_t recv_count = count;
+    size_t send_count = count;
+    size_t recvd      = 0;
+    request *rreq, *sreq;
+
+    if (truncated) {
+        recv_count /= 2;
+    }
+
+    if (extended) {
+        send_count /= 2;
+    }
+
+    if (expected) {
+        rreq = recv_nb(recvbuf, recv_count, recv_dt, RECV_TAG, RECV_MASK);
+        sreq = send_nb(sendbuf, send_count, send_dt, SENDER_TAG);
+    } else {
+        sreq = send_nb(sendbuf, send_count, send_dt, SENDER_TAG);
+
+        wait_for_unexpected_msg(receiver().worker(), 10.0);
+
+        rreq = recv_nb(recvbuf, recv_count, recv_dt, RECV_TAG, RECV_MASK);
+    }
+
+    /* progress both sender and receiver */
+    wait(rreq);
+    if (sreq != NULL) {
+        wait(sreq);
+        request_release(sreq);
+    }
+
+    recvd = rreq->info.length;
+    if (!truncated) {
+        EXPECT_UCS_OK(rreq->status);
+        EXPECT_EQ((ucp_tag_t)SENDER_TAG, rreq->info.sender_tag);
+    } else {
+        EXPECT_EQ(UCS_ERR_MESSAGE_TRUNCATED, rreq->status);
+    }
+
+    request_release(rreq);
+    return recvd;
+};
+
+UCS_TEST_P(test_ucp_tag_mem_type, basic)
+{
+    ucp_datatype_t type = ucp_dt_make_contig(1);
+
+    ucs::detail::message_stream ms("INFO");
+    ms <<"TEST: " << ucs_memory_type_names[m_send_mem_type] << " <-> "
+                  << ucs_memory_type_names[m_recv_mem_type];
+
+    for (unsigned i = 1; i <= 7; ++i) {
+        size_t max = (long)pow(10.0, i);
+        size_t length = ucs::rand() % max + 1;
+
+        m_recv_mem_buf = new mem_buffer(length, m_recv_mem_type);
```

Comment:
btw, what about do not allocate it dynamically? so, can just do (remove `m_recv_mem_buf` and `m_send_mem_buf` fields):
```
mem_buffer recv_mem_buf(length, m_recv_mem_type);
mem_buffer send_mem_buf(length, m_send_mem_type);
```